### PR TITLE
feature/session-object/partial-fill-orders/ioc-order-flag/balance-not-enough-fix

### LINKED ISSUE #54: Handling of partially filled orders.  "Not Enough balance" errors on sell.
https://github.com/CyberPunkMetalHead/gateio-crypto-trading-bot-binance-announcements-new-coins/blob/master/main.py#L223

At line 223, we assume that the order has been filled.  We write to the order.json file the full amount in this assumption.
At this point, the order may be in a status of `open`, `cancelled`, `filled` or `closed`

One change I have made is:
1) to change the order create to use the `time_in_force=ioc` flag.  Immediate or cancel.  This makes my order automatically cancel if any or none is filled.  The status is 'cancelled'.  I don't ever want to leave an 'open' status order.

Here is my definition of my place_order method.  Notice the `ioc` provision.

```
def place_order(base,quote, amount, side, last_price):
    """
    Args:
    'DOT', 'USDT', 50, 'buy', 400
    """
    try:
        order = Order(amount=str(float(amount)/float(last_price)), price=last_price, side=side, currency_pair=f'{base}_{quote}', time_in_force='ioc')
        order = spot_api.create_order(order)
        t = order
        logger.info(f"PLACE ORDER: {t.side} | {t.id} | {t.account} | {t.type} | {t.currency_pair} | {t.status} | amount={t.amount} | price={t.price} | left={t.left} | filled_total={t.filled_total} | fill_price={t.fill_price}")
    except Exception as e:
        logger.error(e)
        raise

    else:
        return order
```
        

The second and important change I made is to:

2) Update the buy amount in the order.json file to be accurate to what your order came back with.

At line 223, I have changed the code to this

```
else:
                        if test_mode:
                            order_status = order[announcement_coin]['status']
                        else:
                            order_status = order[announcement_coin]['_status']

                        message = f'Order created on {announcement_coin} at a price of {price} each.  {order_status=}'
                        logger.info(message)
                        

                        if order_status == 'filled' or order_status == "closed":
                            if test_mode and float(order[announcement_coin]['_left']) > 0 and float(order[announcement_coin]['_amount']) > float(order[announcement_coin]['_left']):
                                # you can only sell what you have. Minus fees.  Look for unfulfilled
                                newAmount = float(order[announcement_coin]['_amount']) - float(order[announcement_coin]['_left']) - float(order[announcement_coin]['_fee'])
                                order[announcement_coin]['volume'] = newAmount
                            else:
                                store_order('order_fulfilled.json', order)

                                # you can only sell what you have. Minus fees.  Look for unfulfilled
                                newAmount = float(order[announcement_coin]['_amount']) - float(order[announcement_coin]['_left']) - float(order[announcement_coin]['_fee'])
                                order[announcement_coin]['_amount'] = newAmount

                            store_order('order.json', order)
                            
                            if not test_mode and enable_sms:
                                try:
                                    send_sms_message(message)
                                except Exception:
                                    pass
                        elif order_status == 'open' or order_status == 'cancelled':
                            if not test_mode and order_status == 'open':
                                # cancel orders and try again in the next iteration
                                cancel_open_order(order[announcement_coin]['_id'], announcement_coin, pairing)
                                logger.info(f"Cancelled order {order[announcement_coin]['_id']} .  Waiting for status of 'filled/closed' for {announcement_coin}")
                            
                            order.clear()  # reset for next iteration
                            
```

You must deflate the order by the fee amount too to make sure you get a successful sell order filled.

### PR DESCRIPTION
**Pull request by busyuqboy**
# Partial orders / "BALANCE_NOT_ENOUGH" error

## Summary
*Resolves #54*
- NOT_ENOUGH_BALANCE error on sell
- Handling paritally filled orders (by keeping a session object summing amounts and fees across multiple orders)
- Using ioc to fill orders (or cancel immediately)

## Change orders to `ioc` flag
+ [x] Change orders to use `time_in_force='ioc'` flag which forces orders to be either `closed` (totally filled) or `cancelled` (none filled or paritally filled.  Both buy and sell side orders will use this new flag by default.
Reasons:
- Stops orders from being created with an `open` status since the asking price was passed over/missed in the volatility of the trading.  Worst case, you can buy in successfully and then leave a sell order in an `open` state.  
- Using `immediate or cancel` doesn't require a POST request to cancel the open order each main thread iteration.  Any order that the status doesn't come back `closed` will be automatically cancelled.  allowing an immediate place order at the next iteration.

## Reslove `BALANCE_NOT_ENOUGH` issue
+ [x] You must deflate the sell amount by the fees spent or you will get a 400 error of `BALANCE_NOT_ENOUGH`. 
- this error happens on the sell order.
- simply put, you need to subtract the fee used to close on the amount you gained.

## Add `session` object / resolves multiple orders due to partial fills
+ [x] a new `session` dictionary variable to contain the many possible buy and sell orders the process may take.
+ [x] at each iteration, the `session` object will keep a running total of the `total_volume`, `total_amount` and `total_fees` you have collected for as many orders as required.
+ [x] store session data into `session.json` at the end of the buy phase and at the end of the sell phase for analysis of transactions
+ [x] updated test mode to perform a partial fill by default.
+ [ ] update test mode to perform a partial fill by default.  (not ready)

### CODE REVIEW

File: config.example.yml [hunk 1]
```diff
@@ -12,7 +12,8 @@
     ENABLE_TSL: True
     TSL: -4
     TTP: 2
-  LOGGING:
+   LOGGING:
```
Comment (cryptmjt) [inline config.example.yml:15]: Is this extra space on purpose? :)

Comment (busyuqboy) [reply config.example.yml:15]: No, not on purpose.
